proposals/NNNN-discordbm.md

```diff
@@ -0,0 +1,263 @@
+
+/// Get an `AsyncStream` of `Gateway.Event`s
+let stream = await bot.makeEventsStream()
+
+/// Handle each event in the stream
+/// When used in the @main entry point, this stream will never end,
+/// therefore preventing your executable from exiting
+for await event in stream {
+    switch event.data {
+    case let .messageCreate(message):
+        print("NEW MESSAGE!", message)
+
+        /// You can also use the `GatewayEventHandler` protocol for more convenience
+    default: break
+    }
+}
+```
+
+This is also in line with the latest structured concurrency advancements, and users won't need to block a thread or call functions such as `RunLoop.main.run()`. Generally speaking, DiscordBM has gone full-in on structured concurrency and makes extensive use of actors. The whole library is already in full compliance with `-strict-concurrency=complete`.
+
+`GatewayManager` also needs to keep the Discord connection alive. This happens in two ways.
+
+1. The usual websocket ping-pongs.
+2. By reconnecting when Discord closes a Websocket connection.
+
+Connection closures are normal and initiated by Discord when they see fit. The closures could happen 10+ times a day, but are mostly not distruptive. Upon a conneciton closure, DiscordBM will correctly resume the last connection, and Discord will send the events lost when there was no connection.
+
+After receiving an event, users can use the `DiscordClient` to communicate with Discord through the REST API. For example you can send a message back to the Discord channel that a message came from, like so:
+
+```swift
+/// Use `bot.client` to send requests to Discord
+case let .messageCreate(message):
+    let response = try await bot.client.createMessage(
+        channelId: message.channel_id,
+        payload: .init(content: "Got a message: '\(message.content)'")
+    )
+```
+
+And you can safely and effortlessly decode Discord's response to the correct type:
+
+```swift
+/// `message` will be of type `DiscordChannel.Message`
+let message = try response.decode()
+```
+
+The REST API endpoints are all declared as extensions to the `DiscordClient` protocol like so:
+
+```swift
+/// https://discord.com/developers/docs/resources/channel#create-message
+@inlinable
+func createMessage(
+    channelId: ChannelSnowflake,
+    payload: Payloads.CreateMessage
+) async throws -> DiscordClientResponse<DiscordChannel.Message> {
+    let endpoint = APIEndpoint.createMessage(channelId: channelId)
+    return try await self.sendMultipart(
+        request: .init(to: endpoint),
+        payload: payload
+    )
+}
+```
+
+This function returns a `DiscordClientResponse`. The generic `C` type that `DiscordClientResponse` takes makes it possible to decode Discord's responses to a library-defined type as you saw in `let message = try response.decode()`:
+
+```swift
+/// Represents a Discord HTTP response for endpoints that return some data in the body.
+public struct DiscordClientResponse<C>: Sendable where C: Codable {
```

Comment:
Doesn't C also need to be Sendable?

Reply:
Not really. It would be fine for the library to add that constraint, but this hasn't been necessary.

What happens is when you have a value of type `DiscordClientResponse<C>`, you can get a `C` value with the `.decode()` function, and what `.decode()` function returns is always Sendable enforced by e.g. the `DiscordClient.send()` functions which return the `DiscordClientResponse<C>`.

I'm not against adding that, but it won't have too much of a use-case. The decoded value doesn't really _need_ to be Sendable as far as a decoded-object is concerned.